### PR TITLE
[DOC] Created a adding_typehints.md in developers_guide that Fixes issue #1857

### LINKED ISSUE #1857: [DOC] Write developer documentation for type hints including soft dependencies.
### Describe the issue linked to the documentation

Developer documentation does not sufficiently include guidance for adding type hints, espceically in situations where soft dependencies are required.

### Suggest a potential alternative/fix

Add a new page in `docs/developer_guide/` on adding type hints.  [`PyODAdapter`](https://github.com/aeon-toolkit/aeon/blob/main/aeon/anomaly_detection/_pyodadapter.py) is a good example of type hints with soft dependencies, following [PEP 563](https://peps.python.org/pep-0563/).

### PR DESCRIPTION
**Pull request by vedpawar2254**
Added a documentation file about adding type hints

#### Reference Issues/PRs

Fixes  #1857

#### What does this implement/fix? Explain your changes.

This adds a file to the developers_guide and the file contains a guide to adding type hints in the project

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

### CODE REVIEW

**Comment by baraline** [comment]
Hi! Thanks for the PR, I'm wondering if that kind of content shouldn't be added as a section in the coding standard page rather than on its own. 
@MatthewMiddlehurst do we consider type hint as a standard or simply a good practice?

File: docs/developer_guide/adding_typehints.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+Type hints are a way to indicate the expected data types of variables, function parameters, and return values in Python. They enhance code readability and help with static type checking, making it easier to catch errors before runtime. For more information, refer to [PEP 563](https://www.python.org/dev/peps/pep-0563/) which discusses forward references.
+
+
+Type hints act as a form of documentation that helps developers understand the types of arguments a function expects and what it returns.
+
+Example:
+
+
+You can provide type hints for function parameters and return values. This helps other developers understand what types of arguments are expected by the function and what type the function returns.
```
Comment (MatthewMiddlehurst) [inline docs/developer_guide/adding_typehints.md:13]: There is a bit a repetition here.

**Review by MatthewMiddlehurst — CHANGES_REQUESTED** [review]

**Comment by MatthewMiddlehurst** [comment]
> Hi! Thanks for the PR, I'm wondering if that kind of content shouldn't be added as a section in the coding standard page rather than on its own. @MatthewMiddlehurst do we consider type hint as a standard or simply a good practice?

I think see where this ends up then decide

**Comment by krushndayshmookh** [comment]
I think this should be enforced as a standard to ensure consistent typing as tooling support is important to ensure code quality.

**Review by MatthewMiddlehurst — APPROVED** [review]

**Pull request merged**